### ISSUE TITLE
cloning the data is a breaking change

### ISSUE THREAD

**Issue by olsonpm** [clay/clay-log#21]
[PR #20](https://github.com/clay/clay-log/pull/20) introduced a breaking change by cloning the data passed into the logger.

In our code this resulted in Errors being removed from our logs e.g.

```js
catch (err) {
  log('error', 'there was an error', err);
}
```

since `_.isEqual(cloneDeep(err), {}) === true`

the problematic code specifically is [line 177](https://github.com/clay/clay-log/pull/20/commits/1344862348cecde15b3be268f59d785a89084361#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R177)

Object.assign would result in the same behavior

**Comment by mrollinsiv** [comment]
@mattoberle 

A new test case should be added to the unit tests. The error appears to be that non-enumarable properties are dropped when calling `cloneDeep`. In the case above, we appear to have lost the message/stack on the error.

**Comment by mattoberle** [comment]
I haven't personally encountered this, but that may be because we typically do one of these two options for reporting errors:

```js
log(err);
```

or

```js
log('error', 'my message', { stack: err.stack });
```

I'm trying to think about how we can account for the non-enumerable properties while still preventing the mutation side effects that `clay-log` could otherwise cause.

The third argument of `clay-log` is intended to be a JSON-serializable object containing additional context. The first argument is the one with special handling for instances of `Error`. 

Maybe we could extend that handling for the third argument as well?
That would mean extracting `.stack`, `.message` and (`.name`?).

```js
data = data instanceof Error
  ? { stack: data.stack, message: data.message, name: data.name }
  : cloneDeep(data);
```

**Comment by mrollinsiv** [comment]
You would still need to cloneDeep it as there may be custom properties on a custom Error. So cloneDeep it and then specifically reset those 3 if it's an error?

**Comment by olsonpm** [comment]
Only the _label is being added correct ?  i.e. that's the only side effect ?

**Comment by olsonpm** [comment]
this problem isn't specific to errors, so creating a one-off solution for it I feel is insufficient.

The api can't change from mutable to non-mutable without there being a breaking change i.e. a major version bump

**Comment by mattoberle** [comment]
@olsonpm you are right about both it being a breaking change and the Error solution not being sufficient (just wrote tests to verify).

I do think that #20 is important when paired with #16 which allows manipulation of the `data` argument for the purpose of providing additional context (or formatting values in a preferred way).

I could publish a `v1.5.3` with the reverted behavior (removing `cloneDeep`) along with a `v2.0.0` that re-introduces the safer behavior, although that might be a little strange to have a 'broken' `v1.5.2` sitting around.

**Comment by olsonpm** [comment]
a broken v1.5.2 is just how it goes.  It'd be stranger if code in that version were to change

please publish a v1.5.3 with the revert like you mention.  That would be great.

**Comment by mattoberle** [comment]
This was fixed in [v1.5.3](https://github.com/clay/clay-log/releases/tag/v1.5.3) which is now published on NPM.

Thank you for raising this issue.
We'll explore options for protecting against mutation and determine how to proceed on that front.
